dev/NumberBox/NumberBox.cpp

```diff
@@ -158,6 +160,20 @@ void NumberBox::OnApplyTemplate()
             }
 
             m_textBoxKeyUpRevoker = textBox.KeyUp(winrt::auto_revoke, { this, &NumberBox::OnNumberBoxKeyUp });
+
+            // Listen to NumberBox::CornerRadius changes so we can change the corner radius of the textbox in code-behind.
+            // Based on corner radius tests, we cannot do this in Visual States as we need to filter some specific corner radii
+            // for the texbox based on the SpinButtonPlacementMode. For example, with SpinButtonPlacementMode::Inline, the T-rule
+            // applies and we should only set the TopLeft and BottomLeft corner radii of the textbox to values > 0. In all other
+            // SpinButtonPlacementModes, we apply NumberBox's corner radius to all textbox corners. Visual States, however, only
+            // set the property when entering the state. Since NumberBox's corner radius can change while in a Visual State
+            // this could potentially lead to incorrect textbox corner radii updates (rounding all four corners with inlined spin buttons
+            // -> breaking T-rule)
```

Comment:
I'm not sure I understand the issue here, are you saying there is no way to set up a template binding within a VSM setter?

`contract7Present:CornerRadius="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=CornerRadius, Converter={StaticResource RightCornerRadiusFilterConverter}}"` doesn't this do it?

Reply:
@StephenLPeters The issue I was seeing (using the CornerRadius test page of the MUXControlsTestApp) was the following: Using template binding in the style (as you have written above), initially, the corner radius was filtered correctly for the textbox in SpinButtonPlacement.Inline mode. However, as soo as as I used the slider to change the *ControlCornerRadius* value used to validate this addition, the new value corner radius values were consistently applied to *all four* edges of NumberBox' TextBox. In other words, the filtering set in the VisualState *SpinButtonsVisible* was no longer being used, causing the T-rule to be broken. 

I then read the [VisualState.Setters](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.visualstate.setters) documentation and it is my current understanding that the filter set in the visual state is only applied on entering the state. If I assign new CornerRadius values *after* entering this state - which is what I was doing by moving the ControlCornerRadius slider - the value set by the visual state (the filter) would no longer apply.

As such, this lead me to essentially writing this lig in code-behind as you can see. If there is a way to get this working with Visual States alone, I'm all ears. It would certainly be a nicer solution 🙂

Reply:
@StephenLPeters I replaced T-rule enforcement with a GoToState(...,L"SpinButtonsAvailable",..) call so all the corner radius code is back in the control template again. Commit with these changes not yet uploaded.

Reply:
@ranjeshj is there someone with better VSM/Binding knowledge than me that could weigh in here? I don't understand why the binding is behaving as Felix-Dev is describing. Its almost as if the TemplateBinding for that property is still active...  @Felix-Dev could you maybe try this...

Reply:
@StephenLPeters It can take a little while longer until I can report back...issue #2221 is really driving me crazy here. I'm wasting soooo much time re-building the entire project that I need a break currently (will have to switch over to the InnerLoop for future PRs).

Reply:
That does sound frustrating :/... Thanks for the awesome work despite that issue!

Reply:
@StephenLPeters Please correct me if I missunderstood your suggestion but here's what I did (only showing the relevant code).

I first created a named binding as a static resource in the template (root grid):
```xml
<Grid.Resources>
    <Binding x:Name="TextBoxCornerRadiusBinding" Path="CornerRadius" RelativeSource="{RelativeSource Mode=TemplatedParent}"/>
</Grid.Resources>
```

In the VSM's setter, I targeted the named binding's converter property:
```xml
<Setter Target="TextBoxCornerRadiusBinding.Converter" Value="{StaticResource LeftCornerRadiusFilterConverter}" />
```

Finally, I replaced the TextBox's template binding with:
```xml
CornerRadius="{StaticResource TextBoxCornerRadiusBinding}"
```

Running this code, I get an unhandled exceptio, with the error text: `Failed to assign to property 'Windows.UI.Xaml.Controls.Control.CornerRadius' because the type 'Windows.UI.Xaml.Data.RelativeSource' cannot be assigned to the type 'Windows.UI.Xaml.CornerRadius'.`

Reply:
I've also created a gif showing the issue we are trying to solve. I'm using template binding without the VisualStateManager.GoToState() call here:

![numberbox_cr_issue](https://user-images.githubusercontent.com/1398851/78918532-d8d5f000-7a90-11ea-9100-66fc99851f7a.gif)

Note how initially the VSM's setter is working and the T-rule is respected (only the left corners of the textbox are rounded, not the corners adjacent to the spin buttons). However, as I change the corner radius, the updated values apply to all *four* corners of the textbox, breaking the T-rule. This is the reason why I introduced the VisualStateManager.GoToState(..., L"SpinButtonsAvailable", ...) call above.

Reply:
I see, maybe this isn't supported in xaml markup then...